ScopedStorage/app/src/main/java/com/samples/storage/mediastore/AddDocumentViewModel.kt

```diff
@@ -0,0 +1,413 @@
+                null,
+                null,
+                null
+            ) ?: return@withContext null
+
+            cursor.use {
+                if (!cursor.moveToFirst()) {
+                    return@withContext null
+                }
+
+                return@withContext cursor.getString(cursor.getColumnIndexOrThrow(FileColumns.DATA))
+            }
+        }
+    }
+
+    /**
+     * Get file details on Api 21
+     *
+     * It uses the classic java.io APIs but can't get the added time as there's not a reliable way
+     * to do so until Api 26
+     */
+    private suspend fun getFileDetailsIoApi21(path: String): FileEntry? {
+        return withContext(Dispatchers.IO) {
+            val file = File(path)
+
+            if (!file.exists()) {
+                return@withContext null
+            }
+
+            return@withContext FileEntry(
+                filename = file.name,
+                size = file.length(),
+                mimeType = URLConnection.guessContentTypeFromName(file.name),
+                // There are no reliable ways to get the added time on Api 26
+                addedAt = -1,
+                path = path
+            )
+        }
+    }
+
+    /**
+     * Get file details on Api 21
+     *
+     * It uses MediaStore to all the file properties
+     */
+    private suspend fun getFileDetails(uri: Uri): FileEntry? {
+        return withContext(Dispatchers.IO) {
+            val cursor = context.contentResolver.query(
+                uri,
+                arrayOf(
+                    FileColumns.DISPLAY_NAME,
+                    FileColumns.SIZE,
+                    FileColumns.MIME_TYPE,
+                    FileColumns.DATE_ADDED,
+                    FileColumns.DATA
+                ),
+                null,
+                null,
+                null
+            ) ?: return@withContext null
+
+            cursor.use {
+                if (!cursor.moveToFirst()) {
+                    return@withContext null
+                }
+
+                val displayNameColumn = cursor.getColumnIndexOrThrow(FileColumns.DISPLAY_NAME)
+                val sizeColumn = cursor.getColumnIndexOrThrow(FileColumns.SIZE)
+                val mimeTypeColumn = cursor.getColumnIndexOrThrow(FileColumns.MIME_TYPE)
+                val dateAddedColumn = cursor.getColumnIndexOrThrow(FileColumns.DATE_ADDED)
+                val dataColumn = cursor.getColumnIndexOrThrow(FileColumns.DATA)
+
+                return@withContext FileEntry(
+                    filename = cursor.getString(displayNameColumn),
+                    size = cursor.getLong(sizeColumn),
+                    mimeType = cursor.getString(mimeTypeColumn),
+                    // FileColumns.DATE_ADDED is in seconds, not milliseconds
+                    addedAt = cursor.getLong(dateAddedColumn) * 1000,
+                    path = cursor.getString(dataColumn),
+                )
+            }
+        }
+    }
+}
+
+data class FileEntry(
```

Comment:
Why not use `@Parcelize` instead of the to/from bundle? -- https://developer.android.com/kotlin/parcelize